sycl/source/detail/scheduler/graph_builder.cpp

```diff
@@ -227,15 +234,21 @@ MemObjRecord *Scheduler::GraphBuilder::getOrInsertMemObjRecord(
 
 void Scheduler::GraphBuilder::updateLeaves(const std::set<Command *> &Cmds,
                                            MemObjRecord *Record,
-                                           access::mode AccessMode) {
+                                           access::mode AccessMode,
+                                           std::vector<Command *> &ToCleanUp) {
 
   const bool ReadOnlyReq = AccessMode == access::mode::read;
   if (ReadOnlyReq)
     return;
 
   for (Command *Cmd : Cmds) {
+    bool WasLeaf = Cmd->MLeafCounter > 0;
     Cmd->MLeafCounter -= Record->MReadLeaves.remove(Cmd);
     Cmd->MLeafCounter -= Record->MWriteLeaves.remove(Cmd);
+    if (WasLeaf && Cmd->MLeafCounter == 0 && Cmd->isSuccessfullyEnqueued() &&
```

Comment:
I'm a little bit lost.
Wouldn't it be simpler to:
1. Add `readyForCleanup` method to Command which do all the required checks for a command
2. Add a "scheduler global" vector `ToCleanUp` which is guarded by a mutex. (can be later converted to thread local if needed)
3.  In `GraphProcessor::enqueue`, check if a dep of a command is `readyForCleanup`(1.) and, if yes, add it to `ToCleanUp` vector. (Can have local cache if needed to reduce number of locks for accessing the vector).
4. In all* Scheduler high level entry points check if we have something in vector and run cleanup.

?

Reply:
I mainly wanted to keep the number of cleanup vector mutex locks to a minimum: currently either the graph or deferred cleanup vector mutex is locked once per attempted cleanup. I haven't considered that if needed, the same can be achieved by using a thread local member variable in scheduler. Also, if we use a mutex instead of a thread local variable, we still cannot minimize its locking (per 1 Scheduler entry point) with a local cache on the `GraphProcessor::enqueue` level. So I would prefer the current solution or a thread local member variable out of those three options.

I currently make the check inside `Command::enqueue` and populate the vector there since the function only reports that the command has been enqueued at some point rather than as part of this specific `Command::enqueue` call. We could return the second piece of information from `Command::enqueue` to move this logic to `GraphProcessor::enqueue`, but considering that that info is only needed for cleanup, I don't think it would be a significant improvement.

So, do you think that switching to a thread local member variable in scheduler and populating it in `Command::enqueue` and `GraphBuilder::updateLeaves` makes sense?

Reply:
I don't think using thread local vars here is anything better. While it might end up less code, I don't think maintaining it will be easy.

Reply:
I tend to agree. I suggest keeping the current solution for now then, unless any other reviewers disagree with that. @romanovvlad Feel free to continue the review post-merge so that we can revisit this topic if needed.